### PR TITLE
[server] Make PrebuildManager use WorkspaceService.startWorkspace (to remove code duplication)

### PR DESCRIPTION
**Pull request by geropl**
## Description

Make PrebuildManager use WorkspaceService instead of WorkspaceStarter directly

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at c6ee7c6</samp>

This pull request improves the code quality and logic of the server component that handles starting workspaces. It removes unused code, refactors common types and methods, and adds some checks for workspace instances.

</details>

## Related Issue(s)
Related: EXP-549

## How to test
 - create a project with prebuild configured
 - see that prebuild gets triggered and executes fine

#### Preview status
🏷️ Name - gpl-ws-starter-3
🔗 URL - [gpl-ws-starter-3.preview.gitpod-dev.com/workspaces](https://gpl-ws-starter-3.preview.gitpod-dev.com/workspaces).
📚 Documentation - See our [internal documentation](https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39) for information on how to interact with your preview environment.
📦 Version - gpl-ws-starter-3-gha.16386
🗒️ Logs - [GCP Logs Explorer](https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-gpl-ws-starter-3%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev)

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/workspace/workspace-service.ts [hunk 1]
```diff
@@ -437,10 +437,21 @@ export class WorkspaceService {
         user: User,
         workspaceId: string,
         options: StartWorkspaceOptions = {},
+        restrictToRegular = true,
     ): Promise<StartWorkspaceResult> {
         await this.auth.checkPermissionOnWorkspace(user.id, "start", workspaceId);
 
-        const workspace = await this.doGetWorkspace(user.id, workspaceId);
+        const { workspace, latestInstance } = await this.getWorkspace(user.id, workspaceId);
```
Comment (geropl) [inline components/server/src/workspace/workspace-service.ts:444]: This was the only missing functionality compared to GitpodServerImpl.startWorkspace, which was missed during earlier refactorings.

File: components/server/src/workspace/workspace-service.ts [hunk 2]
```diff
@@ -437,10 +437,21 @@ export class WorkspaceService {
...
+        if (latestInstance) {
+            if (latestInstance.status.phase !== "stopped") {
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-service.ts:446]: @geropl I did not understand this change. It seems to change semantic?

Comment (geropl) [reply components/server/src/workspace/workspace-service.ts:446]: I tried to explain [here](https://github.com/gitpod-io/gitpod/pull/18653#discussion_r1314620049): This change is here to make this method en-par with GitpodServerImpl.startWorkspace.

It's not relevant for Prebuilds, bc. Prebuilds are always `Workspace` 1 <> 1 `WorkspaceInstance`.

File: components/server/src/workspace/workspace-service.ts [hunk 3]
```diff
@@ -437,10 +437,21 @@ export class WorkspaceService {
         user: User,
         workspaceId: string,
         options: StartWorkspaceOptions = {},
+        restrictToRegular = true,
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-service.ts:440]: (nit) Should it be part of StartWorkspaceOptions?

File: components/server/src/workspace/workspace-service.ts [hunk 4]
```diff
@@ -456,7 +467,7 @@ export class WorkspaceService {
             };
         }
 
-        if (workspace.type !== "regular") {
+        if (restrictToRegular && workspace.type !== "regular") {
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-service.ts:470]: I wonder whether here we can check whether it is first start or restart instead? and only throw if we are trying to restart a non regular workspace?

**Review by akosyakov — APPROVED** [review]
I smoked test workspace starts. It still works. I wanted to check prebuilds, but I messed up preview env as it seems. Getting not disk space.

File: components/server/src/workspace/workspace-service.ts (hunk 4 above)
Comment (geropl) [reply components/server/src/workspace/workspace-service.ts:470]: Note that the API does not allow to start any non-regular workspace. And we already do have a check for "is this a restart?" in WorkspaceStarter.startWorkspace. I agree that we eventually should move that check up here.

For now, I opted to keep the existing behavior as close to GitpodServerImpl as possible (especially in my current context, where I did not intend to touch it originally).

**Comment by geropl** [comment]
> I wanted to check prebuilds, but I messed up preview env as it seems. Getting not disk space.

Ough. Tested that yesterday, so should be fine. :+1:

File: components/server/src/workspace/workspace-service.ts (hunk 3 above)
Comment (geropl) [reply components/server/src/workspace/workspace-service.ts:440]: I hoped to get rid of this eventually, and that's why I tried to keep it separate. But maybe it should have gone in there anyway... :+1:

**Pull request merged**